OPHD/StructureComponent.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include <NAS2D/Utility.h>
+
+class Structure;
+class StructureManager;
+class StructureComponent;
+
+
+/**
+ * Key type for identifying a specific structure instance.
+ * The key for any given structure is guaranteed to remain unchanged for the lifetime of the structure.
+ * The key for any given structure is guaranteed to be unique during the lifetime of the structure.
+ *
+ * Every structure has a Structure instance. The Structure pointer is used as key to allow O(1) access
+ * to the Structure instance. This is an internal detail and should not be relied upon by code handling the key.
+ */
+class SKey
+{
+private:
+	Structure* mStructure;
+public:
+	SKey(Structure* structure) : mStructure(structure) {}
+
+	/** Comparison operators to allow using this type in ordered containers such as maps and sets. */
+	bool operator<(const SKey& rhs) const { return mStructure < rhs.mStructure; }
+
+	/** Do not call this function directly. It is intended only for GetComponent/TryGetComponent. */
+	Structure* getInternal() { return mStructure; }
+};
```

Comment:
The reason I added this class was because [Leeor was hesitant](https://github.com/OutpostUniverse/OPHD/issues/824#issuecomment-797144678) to directly using the pointer as key. I explained the purpose of this class [here](https://github.com/OutpostUniverse/OPHD/issues/824#issuecomment-798878530) - encapsulating the key in a class to abstract away what the key actually is was to remedy Leeor's concerns. The idea is that code would do GetComponent<Structure>(key), and thus be agnostic to whether the key is a Structure* or something else. Using a typedef or overloading -> won't accomplish that abstraction, since you'd be tempted to just assume it's a Structure* when working with the key. That's ugly too unless the Structure class itself eventually is converted to a StructureComponent too.

If you don't think abstraction is warranted, then we should just use Structure* directly, like I did in my fork. Do the two of you have a common view on this? I don't want to go back and forth on this several times.

Reply:
That makes a little more sense now that I've re-read stuff.

Admittedly the `typedef`/`using` alias does expose the underlying type. That would allow someone to make assumptions that would later bite them if the alias was changed. Though assuming they stuck to convention of using the type as a handle, it would be pretty easy to change just the alias definition line to change between pointer or index representation. I view it a bit like Windows `HANDLE` objects, which are effectively pointers, though it would be very questionable to try and dereference them and directly read the data in the structures they point to.

I'd say just leave it for now. I think I was more looking for an explanation.